File: neural_network_python/dense_layer.py
from __future__ import annotations
from utils import ActFunc
import utils

class DenseLayer:
    
    # ------------------------------------------------------------------------------
    def __init__(self, numNodes: int, numWeightsPerNode: int, actFunc: ActFunc = ActFunc.Relu) -> None:
        self._output = []
        self._error = []
        self._bias = []
        self._weights = []
        self._actFunc = actFunc
        self._initNodes(numNodes, numWeightsPerNode)

    # ------------------------------------------------------------------------------
    def output(self) -> tuple[float]:
        return tuple(self._output)
    
    # ------------------------------------------------------------------------------
    def error(self) -> tuple[float]:
        return tuple(self._error)
    
    # ------------------------------------------------------------------------------
    def bias(self) -> tuple[float]:
        return tuple(self._bias)
    
    # ------------------------------------------------------------------------------
    def weights(self) -> tuple[list[float]]:
        return tuple(self._weights)
    
    # ------------------------------------------------------------------------------
    def actFunc(self) -> ActFunc:
        return self._actFunc
    
    # ------------------------------------------------------------------------------
    def numNodes(self) -> int: 
        return len(self._output)
    
    # ------------------------------------------------------------------------------
    def numWeightsPerNode(self) -> int:
        return len(self._weights[0]) if len(self._weights) > 0 else 0
    
    # ------------------------------------------------------------------------------
    def feedforward(self, input: tuple[float]) -> None:
        for i in range(self.numNodes()):
            value = self._bias[i]
            for j in range(self.numWeightsPerNode()):
                if j < len(input):
                    value += input[j] * self._weights[i][j]
            self._output[i] = utils.actFuncOutput(value, self._actFunc)

    # ------------------------------------------------------------------------------
    def backpropagateOutput(self, output: tuple[float]) -> None:
        for i in range(self.numNodes()):
            if i < len(output):
                error = output[i] - self._output[i]
                self._error[i] = error * utils.actFuncDelta(self._output[i], self._actFunc)

    # ------------------------------------------------------------------------------
    def backpropagateHidden(self, nextLayer: DenseLayer) -> None:
        for i in range(self.numNodes()):
            error = 0.0
            for j in range(nextLayer.numNodes()):
                error += nextLayer.error()[j] * nextLayer.weights()[j][i]
            self._error[i] = error * utils.actFuncDelta(self._output[i], self._actFunc)

    # ------------------------------------------------------------------------------
    def optimize(self, input: tuple[float], learningRate: float) -> None:
        for i in range(self.numNodes()):
            self._bias[i] += self._error[i] * learningRate
            for j in range(self.numWeightsPerNode()):
                if j < len(input):
                    self._weights[i][j] += self._error[i] * learningRate * input[j]

    # ------------------------------------------------------------------------------
    def _initNodes(self, numNodes: int, numWeightsPerNode: int) -> None:
        utils.initList(self._output, numNodes, 0)
        utils.initList(self._error, numNodes, 0)
        utils.initRandomList(self._bias, numNodes)
        utils.initRandom2DList(self._weights, numNodes, numWeightsPerNode)

File: 2023-10-13 - Neurala nätverk i mjukvara (del IV)/neural_network/src/dense_layer.cpp
#include "dense_layer.hpp"

namespace yrgo {
namespace machine_learning {

namespace {

// --------------------------------------------------------------------------------
double GetActFuncOutput(const double sum, const enum ActFunc act_func) {
    return act_func == ActFunc::kRelu ?
        utils::math::Relu(sum) : utils::math::Tanh(sum);
}

// --------------------------------------------------------------------------------
double GetActFuncDelta(const double output, const enum ActFunc act_func) {
    return act_func == ActFunc::kRelu ?
        utils::math::ReluDelta(output) : utils::math::TanhDelta(output);
}

} /* namespace */

// --------------------------------------------------------------------------------
DenseLayer::DenseLayer(const std::size_t num_nodes,
                       const std::size_t num_weights_per_node,
                       const enum ActFunc act_func) 
    : act_func_{act_func} {
    utils::random::Init();
    output_.resize(num_nodes, 0);
    utils::random::InitVector<double>(bias_, num_nodes, 0, 1);
    error_.resize(num_nodes, 0);
    utils::random::InitVector<double>(weights_, num_nodes, num_weights_per_node, 0, 1);
}

// --------------------------------------------------------------------------------
void DenseLayer::Feedforward(const std::vector<double>& inputs) {
    for (std::size_t i{}; i < NumNodes(); ++i) {
        double sum{ bias_[i] };
        for (std::size_t j{}; j < NumWeightsPerNode() && j < inputs.size(); ++j) {
            sum += inputs[j] * weights_[i][j];
        }
        output_[i] = GetActFuncOutput(sum, act_func_);
    }
}

// --------------------------------------------------------------------------------
void DenseLayer::Backpropagate(const std::vector<double>& reference) {
    for (std::size_t i{}; i < NumNodes() && i < reference.size(); ++i) {
        const double error = reference[i] - output_[i];
        error_[i] = error * GetActFuncDelta(output_[i], act_func_);
    }
}

// --------------------------------------------------------------------------------
void DenseLayer::Backpropagate(const DenseLayer& next_layer) {
    for (std::size_t i{}; i < NumNodes(); ++i) {
        double error{};
        for (std::size_t j{}; j < next_layer.NumNodes(); ++j) {
            error += next_layer.error_[j] * next_layer.weights_[j][i];
        }
        error_[i] = error * GetActFuncDelta(output_[i], act_func_);
    }
}

// --------------------------------------------------------------------------------
void DenseLayer::Optimize(const std::vector<double>& inputs, const double learning_rate) {
    for (std::size_t i{}; i < NumNodes(); ++i) {
        bias_[i] += error_[i] * learning_rate;
        for (std::size_t j{}; j < NumWeightsPerNode() && j < inputs.size(); ++j) {
            weights_[i][j] += error_[i] * learning_rate * inputs[j];
        }
    }
}

} /* namespace machine_learning */
} /* namespace yrgo */
